### ISSUE TITLE
When copy block link ,the editor page close that can't copy block link to another page.

### ISSUE THREAD

**Issue by miludexiaoxigua** [nothingislost/obsidian-hover-editor#86]
- Kanban1.2.37，Copy Block Link1.0.4，Hover Editor0.7.3
- 
https://user-images.githubusercontent.com/63856326/161380108-8f874eac-71dd-44bf-a073-9f1347b47549.mp4

**Comment by miludexiaoxigua** [comment]
- obsidian 0.14.2

**Comment by nothingislost** [comment]
Hmm I can't replicate this when the popover is pinned. What happens if you left click inside the popover a few times before trying to select the text and then do "copy link to block"?

**Comment by miludexiaoxigua** [comment]
- Same result as above.
- In hover editor 0.6.0.3 , It can hover and write multiple pages of text at the same time.  When click the 202204021850kanban，the two pinned hover text will not pinned , it return the 202204021850kanban .

**Comment by pjeby** [comment]
The issue here is that clicking on a context menu item that isn't over the hover editor in an unpinned popover will cause an explicit close (via onMouseDown() in onLinkHover.ts).  It needs to ignore clicks on context menus spawned by the active popover, but it's not really clear how to tell if a menu was triggered in the popover or not.

A further complication is that it isn't even necessary to click on a menu item for this to happen: simply *mousing over* a context menu opened from an editor is enough to close an unpinned popover!

This seems reminiscent of #94, in that as with modals, we want to keep open menus from triggering a close, but it's more complex since we only want to prevent it if the mouse is actually *over* said menu.

**Comment by nothingislost** [comment]
Thee popovers in the example video are pinned though which is what is confusing here.

For the unpinned use case, does it really matter if the mouse is over the menu? It'd pretty pretty difficult to open a menu on some element outside of the unpinned popover. So I would say if a menu is opened and there's an unpinned popover active, it can be assumed that the menu is related to that popover. If it that's not true, it seems fairly harmless to assume.

I would propose something like this:

```js
patchMenu() {
  const plugin = this;
  const uninstaller = around(Menu.prototype, {
    onload(old) {
      return function (...args: unknown[]) {
        plugin.activePopovers.forEach(popover => {
          if (!popover.isPinned) {
            popover.activeMenu = this;
          }
        });
        return old.call(this, ...args);
      };
    },
    onunload(old) {
      return function (...args: unknown[]) {
        plugin.activePopovers.forEach(popover => {
          if (!popover.isPinned) {
            setTimeout(() => {
              if (popover.activeMenu === this) {
                popover.activeMenu = undefined;
              }
            }, 2000);
          }
        });
        return old.call(this, ...args);
      };
    },
  });
  this.register(uninstaller);
}
```

**Comment by pjeby** [comment]
How would this interact with QuickExplorer?  QE opens and closes a lot of menus and has 'em open simultaneously, with popups that are opened *after* the menus.  This looks to me like QE'd end up with stale popovers, unless I added code to clear activeMenu when QE wants to close a popover.  I'm also wondering about the 2s delay, since ISTM that unsetting the activeMenu won't cause the popover to close, it should just make it so that the close countdown starts once the mouse moves.  And ISTM that the unload shouldn't check for isPinned, as otherwise popovers pinned while the menu is open will end up hanging on to a menu reference.  (QE sometimes pins or unpins popovers to support certain functionality such as right-arrowing from an item into its popover.)

It also might simplify things to have the onload function just register() a callback to unset the activeMenu for the target popover (with or without the timeout).  It does mean the menu holds references to the active popovers, but then they're not going to be closed while the menu is active anyway, and it would cut down on the code a bit while avoiding hairy questions like what if the pinning state changes between load and unload -- it's just, "these are the ones we set it on, so let's take it back off on those if it's still there".

Hm.  That brings me another thought, which is that the initial assignment of menu shouldn't take place if the popover in question already has an activeMenu.  QE opens multiple menus at the same time, so the way this is currently written, every unpinned popover will be linked to the topmost menu, then freed as soon as you navigate up the menu hierarchy (i.e. close the most-recently-opened menu).

**Comment by nothingislost** [comment]
So far, the proposed logic seems to play nicely with QE. That was the first thing I tested :) As you mentioned, QE menus initialize prior to any popover being invoked, thus the `Menu.onload` patch never assigns a QE menu to a popover. Feel free to correct me if that's an incorrect assumption but I can't think of any scenario where you could invoke a QE menu while also having an unpinned popover open. The one exception would be triggering QE via a hotkey or the command palette while you had an unpinned popover open but that's pretty niche.

As for the timeout, it's there to allow the user time to move their mouse back inside the hover editor. The context menu is often outside of the hover editor window and when clicking on a context menu item outside of the hover editor window, if you don't add a delay, the hover editor will close after the wait period (600ms by default) which is likely going to be too fast unless the user changes the default delay.

Here's a revised version:

```js
  patchMenu() {
    const plugin = this;
    const uninstaller = around(Menu.prototype, {
      onload(old) {
        return function (...args: unknown[]) {
          plugin.activePopovers.forEach(popover => {
            if (!popover.isPinned && !popover.activeMenu) {
              popover.activeMenu = this;
              this.register(() => {
                setTimeout(() => {
                  popover.activeMenu = undefined;
                }, 2000); // 10 if we wanted to fallback to the default close delay
              });
            }
          });
          return old.call(this, ...args);
        };
      },
    });
    this.register(uninstaller);
  }
```

**Comment by pjeby** [comment]
>  Feel free to correct me if that's an incorrect assumption but I can't think of any scenario where you could invoke a QE menu while also having an unpinned popover open.

QE opens unpinned popovers and tells them the mouse is over them, and if you then right-arrow into a folder it'll want to close that popover and open a new one.  I was afraid that your first version of the code would keep the earlier popover from closing for 2 seconds, but it doesn't.

> The one exception would be triggering QE via a hotkey or the command palette while you had an unpinned popover open but that's pretty niche.

Heh.  I actually do that all the time now, because if I'm working in an unpinned, autofocused, hover popover, I'll then often open QE with a hotkey.  (I almost never open QE by clicking on the actual items.)

Anyway, the new version works great w/QE, even in this "niche" case. :wink:

What it *doesn't* seem to do is fix the original problem -- nor does the older version.  If I right click in a focused, unpinned source editor, then click an item that lies outside the box boundary, it closes the editor instantly.  i.e. the close logic in onLinkHover.ts is still active.

If I fix the onLinkHover to check for `".hover-editor, .menu"`, then I can take the timeout as low as 10ms and the popover will actually remain open until the mouse goes back into it and then out of it again.  But it doesn't work without a timeout, which surprised me.  I think it's probably the core reaping logic having its way there, responding to the menu click after the menu closes and trying to close all the popups.  A delay of 1ms doesn't work so there must be other code setting timeouts somewhere.  Anyway, the point is that the timeout needs to be long enough to keep the editor from closing immediately; but the editor will remain open indefinitely at that point. until you move the cursor into it (and back out) or over another hover target for the same parent.  (At least, that's what I'm seeing here with the .menu addition to the click logic in onLinkHover.ts.)

tl;dr: it needs the .menu fix, but the new version works fine, even with a very short timeout (because the still-unpinned popover stays open until you move the mouse over and off it or its targetEl, or onto another link that shares its hoverParent).

(Unfortunately, it's not clear that this fixes the actual originally-demoed problem, since as you point out, the demo video is with *pinned* popovers!  I tried to repro w/Kanban but still couldn't.  But we can ask OP to re-try this once we have this logic in, as it's possibly the cause even in the demo case.)

**Comment by nothingislost** [comment]
@miludexiaoxigua if you're a user of the BRAT community plugin, please try checking BRAT for beta plugin updates and see if the latest v0.8.1 release resolves the issue you're seeing here

**Comment by nothingislost** [comment]
This seems to be an installer version issue. Please grab and install an updated Obsidian installer from https://obsidian.md/ and see if that helps.

**Comment by pjeby** [comment]
Also, version 0.8.2 should probably work even with an older installer, but upgrading is a good idea nonetheless.

**Comment by miludexiaoxigua** [comment]
> @miludexiaoxigua if you're a user of the BRAT community plugin, please try checking BRAT for beta plugin updates and see if the latest v0.8.1 release resolves the issue you're seeing here

Thank you very much for solving my problem.
